lib/puppetserver/ca/cli.rb

```diff
@@ -0,0 +1,101 @@
+require 'optparse'
+require 'puppetserver/ca/version'
+require 'puppetserver/ca/setup_command'
+require 'puppetserver/ca/logger'
+
+module Puppetserver
+  module Ca
+    class Cli
+      VALID_COMMANDS = {'setup' => SetupCommand}
+
+      def self.run!(cli_args = ARGV, out = STDOUT, err = STDERR)
+        logger = Puppetserver::Ca::Logger.new(:info, out, err)
+        parser, general_options, unparsed = parse_general_inputs(cli_args)
+
+        if general_options['version']
+          logger.inform Puppetserver::Ca::VERSION
+          return 0
+        end
+
+        subcommand = VALID_COMMANDS[unparsed.shift]
+
+        if general_options['help']
+          if subcommand
+            logger.inform subcommand.parser.help
+          else
+            logger.inform parser.help
+          end
+
+          return 0
+        end
+
+        if subcommand
+          command = subcommand.new(logger)
+          input, exit_code = command.parse(unparsed)
+
+          if exit_code
+            return exit_code
+          else
+            return command.run!(input)
+          end
+        else
+          logger.warn parser.help
+          return 1
+        end
+      end
+
+      def self.parse_general_inputs(inputs)
+        parsed = {}
+        general_parser = OptionParser.new do |opts|
+          opts.banner = 'Usage: puppetserver ca <command> [options]'
+          opts.separator ''
+          opts.separator 'Manage the Private Key Infrastructure for'
+          opts.separator 'Puppet Server\'s built-in Certificate Authority'
+          opts.separator ''
+          opts.separator 'Available Sub-Commands:'
+          VALID_COMMANDS.each_pair do |command, cls|
+            opts.separator "    #{command}\t#{cls.description}"
+          end
+          opts.separator ''
+          opts.separator 'General Options:'
+          opts.on('--help', 'Display this general help output') do |help|
+            parsed['help'] = true
+          end
+          opts.on('--version', 'Display the version') do |v|
+            parsed['version'] = true
+          end
+
+          opts.separator ''
+          opts.separator 'Sub-Command Options:'
+          VALID_COMMANDS.each_pair do |command, cls|
+            opts.separator "  #{command}:"
+            subopts = cls.parser.summarize.
+              drop_while{|l| !(l =~ /Options:/) }.
+              select{|l| l =~ /^\s*--/ }.
+              reject{|l| l =~ /--help|--version/ }
```

Comment:
This is gross and should either be cleaned up or removed. I pushed up this wip (besides it being the end of the day) to get folks feelings on whether or not a direction _like_ this is worthwhile or whether we should dumb down the help output and have a simpler implementation.